src/individual_metrics.py

```diff
@@ -0,0 +1,118 @@
+"""
+This module is intended to run after all data has been collected and added.
+
+to the .json file. The purpose is to create a dictionary where the username is.
+
+the key and other calculated metrics are the values.
+"""
+from __future__ import division
+import json_handler
+import data_collection
+
+# from pydriller.domain.commit import ModificationType
+
+
+def get_commit_average(lines, commits):
+    """Find average lines modified per commit."""
+    # Loop through the dictionary and calculate the average lines per commits
+    # formula for average: (added + deleted)/number_of_commits
+    if commits != 0:
+        average = lines / commits
+        return (int)(average)
+    return 0
+
+
+def parse_for_type(name):
+    """Parse through file name and returns its format."""
+    if "." in name:
+        dot_index = name.find(".")
+        return name[dot_index:]
+    return name
+
+
+def get_file_formats(files):
+    """Create a list of unique file formats."""
+    formats = []
+    for file in files:
+        current_format = parse_for_type(file)
+        if current_format not in formats:
+            formats.append(current_format)
+    return formats
+
+
+# NOTE: this function is temporary to test get_individual_metrics considering
+# there is not a main module that connects eerything yet
+def add_raw_data_to_json(path_to_repo, json_file_name):
+    """Check if raw data is in .json file and collects it if not."""
+    # get the data from .json file
+    current_data = json_handler.get_dict_from_json_file(json_file_name)
+    # check if data has not been collected and do so if it was not
+    if "RAW_DATA" not in current_data.keys():
+        print("Raw data has not been collected, collecting it now...")
+        # collects data from data_collection
+        raw_data = {"RAW_DATA": data_collection.collect_commits_hash(path_to_repo)}
+        # Write raw data to .json file
+        json_handler.add_entry(raw_data, json_file_name)
+        print("Data collected")
+
+
+def calculate_individual_metrics(json_file_name):
+    """Retrieve the data from .json file and create a dictionary keyed by user."""
+    current_data = json_handler.get_dict_from_json_file(json_file_name)
+    # creates a hashmap where the key is the authors username
+    data_list = {}
+    # Check if RAW_DATA is in json
+    if "RAW_DATA" in current_data.keys():
```

Comment:
Is "RAW_DATA" a hardcoded value?

Reply:
`"RAW_DATA"` is the key for the commit data obtained by `data_collection`, we can add this to the dictionary under a different more descriptive name if preferred.

Reply:
Can you point me to the file where "RAW_DATA" is actually being added to the JSON file? This looks sketchy. If a key is called "RAW_DATA", I assume there is something that's not raw data in the same json file?

Reply:
When data is collected from using Pydriller in the `data_collection` module, it is stored as a list of dictionaries of all the instance variables of a commit object. This is what I'm referring to as `RAW_DATA`. I think that this information would still be valuable to store in the `.json` file because it offers specific details about each commit. `add_raw_data_to_json()` function checks if the key `RAW_DATA` exists in the `.json` file and calls the data collection method as well as write the data to the file if it was not found. Our `calculate_individual_metrics()` function then accesses the `.json` file and create a dictionary keyed by usernames. After that, it writes that dictionary into the `.json` file under the `individual_metrics` key. I hope this clarifies the process. Please let me know if you find any issues with it or if you have any requests for change.

Reply:
I see. The individual metric dict is actually not being written into the `.json` but just simply being returned here. I think it's also important to inform the data collection/architect team(I'm not sure who's in charge) if having "RAW_DATA" as a key in the original JSON file since other people/functions might not know it. It requires a correct order of execution now since every procedure needs to interact with the JSON file that happened after calculating individual metrics need to know that the JSON is not simply a list of commit data but key-value pairs of "RAW_DATA" and "INDIVIDUAL_METRICS". With that, I don't know if it's a good idea to store the individual metrics back into the same JSON file. Having a separate JSON file just seems way less complicated to me

Reply:
I think you make great points! I agree that having a separate JSON file would be less complicated. I'm not sure how the architects' team plans to approach this, but I've tried to be in contact to understand the structure they propose. We will hopefully be discussing this tomorrow to create a new, more suitable structure for the JSON file(s). I don't expect this PR to be merged tonight and I realize the multitude of issues present. I wanted to get feedback on how we can start fixing the issues.